### ISSUE TITLE
[rllib] Reproduce issue: with the same seed, multiple runs still have different performance

### ISSUE THREAD

**Issue by yangysc** [ray-project/ray#18245]
### What is the problem?

I found multiple runs for PPO still have different performance even we set the same seed.
How can we obtain the exact same result with the same seed?

Current strategy of rllib sets the platform seed as follows:

https://github.com/ray-project/ray/blob/db9b5f142d8c91a872bff00735eb8574dd374b6e/rllib/evaluation/rollout_worker.py#L77-L112

I attached the results of two runs with both `seed=0`

<img width="810" alt="Snipaste_2021-08-31_15-01-50" src="https://user-images.githubusercontent.com/8980981/131458069-d04aa1db-7735-49fc-a39c-e6f911e28246.png">
<img width="810" alt="Snipaste_2021-08-31_15-02-17" src="https://user-images.githubusercontent.com/8980981/131458079-8bdaca34-2fcc-41d3-bf46-957441c17df9.png">

*Ray version and other system information (Python version, TensorFlow version, OS):*
- Ray-dev.2.0  (commit 8136d2912b195a7a6b84f675e18165f873b2138e)
- Pytorch: 1.8.0+cu111
- CUDA driver: NVIDIA-SMI 470.63.01    Driver Version: 470.63.01    CUDA Version: 11.4  
- python：3.8
- ubuntu 20.04
### Reproduction (REQUIRED)
1. install the latest rllib (commit 8136d2912b195a7a6b84f675e18165f873b2138e)
2. run the `cartpole_lstm` example on two gpus
```python
import argparse
import os

from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole
from ray.rllib.utils.test_utils import check_learning_achieved

parser = argparse.ArgumentParser()
parser.add_argument(
    "--run",
    type=str,
    default="PPO",
    help="The RLlib-registered algorithm to use.")
parser.add_argument("--num-cpus", type=int, default=0)
parser.add_argument(
    "--framework",
    choices=["tf", "tf2", "tfe", "torch"],
    default="torch",
    help="The DL framework specifier.")
parser.add_argument("--eager-tracing", action="store_false")
parser.add_argument("--use-prev-action", action="store_true")
parser.add_argument("--use-prev-reward", action="store_true")
parser.add_argument(
    "--as-test",
    action="store_true",
    help="Whether this script should be run as a test: --stop-reward must "
    "be achieved within --stop-timesteps AND --stop-iters.")
parser.add_argument(
    "--stop-iters",
    type=int,
    default=200,
    help="Number of iterations to train.")
parser.add_argument(
    "--stop-timesteps",
    type=int,
    default=100000,
    help="Number of timesteps to train.")
parser.add_argument(
    "--stop-reward",
    type=float,
    default=150.0,
    help="Reward at which we stop training.")

if __name__ == "__main__":
    import ray
    from ray import tune

    args = parser.parse_args()
    cwd_path = os.getcwd()
    print('cwd path', cwd_path)
    logdir = cwd_path + '/log'
    print(logdir)
    if not os.path.isdir(logdir):
        os.mkdir(logdir)
    ray.init(logging_level='DEBUG')

    configs = {
        "PPO": {
            "num_sgd_iter": 5,
            "model": {
                "vf_share_layers": True,
            },
            "vf_loss_coeff": 0.0001,
        },
        "IMPALA": {
            "num_workers": 2,
            "num_gpus": 0,
            "vf_loss_coeff": 0.01,
        },
    }

    config = dict(
        configs[args.run],
        **{
            "env": StatelessCartPole,
            # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
            "num_gpus": 1,
            "num_workers": 1,
            "num_gpus_per_worker": (2 - 1) / (1+1),
 
            "grad_clip": 0.5,#tune.grid_search([1, 5]),

            "seed": 0,
            "model": {
                "use_lstm": True,
                "lstm_cell_size": 256,
                "lstm_use_prev_action": args.use_prev_action,
                "lstm_use_prev_reward": args.use_prev_reward,
            },
            "framework": args.framework,
            # # Run with tracing enabled for tfe/tf2?
            # "eager_tracing": args.eager_tracing,
        })

    stop = {
        "training_iteration": args.stop_iters,
        "timesteps_total": args.stop_timesteps,
        "episode_reward_mean": args.stop_reward,
    }


    results = tune.run(args.run, config=config, verbose=3, checkpoint_freq=50, local_dir=logdir, stop=stop)

    if args.as_test:
        check_learning_achieved(results, args.stop_reward)
    ray.shutdown()


```

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by sven1977** [comment]
Hey @yangysc . There is one more location, where we need to set the seed, which is in Trainer's setup.
This is not done yet in master, but there is another open PR that fixes this issue. I can separate it out and make it its own PR:
https://github.com/ray-project/ray/pull/18115

It's basically just adding the:
```
    @override(Trainable)
    def setup(self, config: PartialTrainerConfigDict):
        _update_global_seed(config, config.get("seed"))

```

to trainer.py (and importing this function from `ray.rllib.evaluation.rollout_worker` (we'll probably move it into the utils dir though).

**Comment by yangysc** [comment]
> Hey @yangysc . There is one more location, where we need to set the seed, which is in Trainer's setup.
> This is not done yet in master, but there is another open PR that fixes this issue. I can separate it out and make it its own PR:
> #18115
> 
> It's basically just adding the:
> 
> ```
>     @override(Trainable)
>     def setup(self, config: PartialTrainerConfigDict):
>         _update_global_seed(config, config.get("seed"))
> ```
> 
> to trainer.py (and importing this function from `ray.rllib.evaluation.rollout_worker` (we'll probably move it into the utils dir though).

Hi, Sven, thanks

 I added this line as you suggested, but the results are still not exactly the same.  Could you obtain the same results for the above `cartpole` example?

 I am not sure if rllib needs to take care of the `dataloader` randomness problem of pytorch (https://pytorch.org/docs/stable/notes/randomness.html).

> DataLoader
DataLoader will reseed workers following Randomness in multi-process data loading algorithm. Use worker_init_fn() and generator to preserve reproducibility:
```
def seed_worker(worker_id):
    worker_seed = torch.initial_seed() % 2**32
    numpy.random.seed(worker_seed)
    random.seed(worker_seed)

g = torch.Generator()
g.manual_seed(0)

DataLoader(
    train_dataset,
    batch_size=batch_size,
    num_workers=num_workers,
    worker_init_fn=seed_worker
    generator=g,
)
 ```
![image](https://user-images.githubusercontent.com/8980981/131595850-c7167306-829f-4665-9ee5-8496f7938f10.png)

**Comment by yangysc** [comment]
Btw, even without any gpu, the results are not the same, either. :<
![image](https://user-images.githubusercontent.com/8980981/131601478-44683f20-2107-4aaa-be2f-a4f51f197f5d.png)
![image](https://user-images.githubusercontent.com/8980981/131601490-16149764-a8ce-42f5-aa69-c22eaa30fb91.png)

**Comment by yangysc** [comment]
Hi, Sven, @sven1977  I updated the new `rollout_worker` code. But still there are some randomnesses.
If we print the `seed` value in `stateless_cartpole.py`, and we use 3 workers (seed is 0) to learn it.
```python
    def seed(self, seed=None):
        print(f"seed is {seed}")

        self.np_random, seed = seeding.np_random(seed)
        return [seed]
```
Two different runs would print different order of the seed values (though the values  are the same, but the order is random).


```
(pid=10724) seed is None
(pid=10724) seed is 2002
(pid=21272) seed is None
(pid=21272) seed is 1001
(pid=17380) seed is None
(pid=17380) seed is 3003

Another run:
(pid=17772) seed is None
(pid=17772) seed is 2002
(pid=10440) seed is None
(pid=10440) seed is 3003
(pid=16848) seed is None
(pid=16848) seed is 1001
```

Maybe this is the reason why there are still some randomnesses?   Coud we ensure the consistent order?  Thanks a lot~ :）

**Comment by gjoliver** [comment]
Are you using any remote worker? What happens if we use a few envs with a single local worker (num_workers=0)?
Different order of random seeds printed feels like an artifact of using async distributed training (logs getting to the head node at non-deterministic time).

**Comment by yangysc** [comment]
> Are you using any remote worker? What happens if we use a few envs with a single local worker (num_workers=0)?


Yes, I used remote workers with `local_mode`  being False. I only tested PPO algorithm with num_workers > 0 since I need it for fast training speed.  


> Different order of random seeds printed feels like an artifact of using async distributed training (logs getting to the head node at non-deterministic time).

Do you mean that only logs are printed randomly, but the actual order of different workers is consistent for different runs with the same seed?

**Comment by gjoliver** [comment]
Yeah. I think that is very likely.

Understand you have num_workers > 0, are these workers actually running on different hosts or on a single computer?
I guess multi-process program is also subject to the same randomness, even if they are all on the same computer.

I need to double check, but I feel like in a distributed setting, it's really hard to make things 100% deterministic. E.g., the samples fetched will arrive at random order, so depending on how we collect sample batches, the batches may not look exactly the same.
And a very small difference in the policy may result in different behaviors in the worker, and you will get completely different results from that point on.

**Comment by sven1977** [comment]
It would be very nice, if we had a `ray.set_deterministic(True)`, though.

**Comment by sven1977** [comment]
For PPO, this may actually work, b/c even with num_workers > 1, the ray.get would return all collected SampleBatches en-sync. For the async algos (APPO, IMPALA, APEX, A3C, etc..), I'm with @gjoliver , this would currently be near impossible to achieve.

**Comment by avnishn** [comment]
UPDATE:
I ran the same experiment with the file `cartpole_lstm.py`


and these were my results when I had 0 additional workers:
https://tensorboard.dev/experiment/ndJTJi6rT868u8dA8ieXsg/

these were my results when I had 1 additional worker which I gave half a gpu to,
while I gave half a gpu to my trainer:
https://tensorboard.dev/experiment/9MRMrADJSzOXYyaVMDbtbA/

I don't see a difference in the performance between the 2 runs in both cases. 

**I'm inclined to close this issue. I can also try reproducing with 2 gpu's but I think there will be the same effect as if
I split the gpu in half -- like I did above.**

Thanks,
@avnishn

**Comment by avnishn** [comment]
Can you try reproducing on colab @yangysc

**Comment by avnishn** [comment]
here is the same experiment (cartpole lstm) but run on 2 gpus, as a redundancy:
https://tensorboard.dev/experiment/vAIGRhNET3K7lKfOXxaWIw/#scalars&_smoothingWeight=0

I don't see a difference in the performance between the 2 runs of the same seed.

**Comment by yangysc** [comment]
> UPDATE: I ran the same experiment with the file `cartpole_lstm.py`
> 
> and these were my results when I had 0 additional workers: https://tensorboard.dev/experiment/ndJTJi6rT868u8dA8ieXsg/
> 
> these were my results when I had 1 additional worker which I gave half a gpu to, while I gave half a gpu to my trainer: https://tensorboard.dev/experiment/9MRMrADJSzOXYyaVMDbtbA/
> 
> I don't see a difference in the performance between the 2 runs in both cases.
> 
> **I'm inclined to close this issue. I can also try reproducing with 2 gpu's but I think there will be the same effect as if I split the gpu in half -- like I did above.**
> 
> Thanks, @avnishn

Hi, @avnishn  Thanks for your experiments.   May I ask what version of ray you tested?  I tested using this wheel then `https://github.com/ray-project/ray/commit/8136d2912b195a7a6b84f675e18165f873b2138e`.  Maybe it is due to the multi-GPU bug that is already fixed by this commit `https://github.com/ray-project/ray/pull/18937`?  I can re-test the one that you used later today.

BTW, I want to make sure that, for `cuda` version > 10.2, does this line below already set all things deterministic, since it bypasses this line `torch.use_deterministic_algorithms(True)`. 
https://github.com/ray-project/ray/blob/fe413c3c5eef400a03a8e5407b295eb122f0b244/rllib/utils/debug.py#L97-L98

Thanks again~

**Comment by avnishn** [comment]
You can share them using tensorboard dev:
https://tensorboard.dev/

What type of compute are you using? Are you using any cloud compute, or is this a personal machine?

**Comment by richardliaw** [comment]
Awesome!
